sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -144,14 +146,22 @@ object DataSourceV2Strategy extends Strategy {
       WriteToDataSourceV2Exec(writer, planLater(query)) :: Nil
 
     case AppendData(r: DataSourceV2Relation, query, _) =>
-      val writeBuilder = r.newWriteBuilder(query.schema)
-      writeBuilder match {
-        case s: SupportsSaveMode =>
-          val write = s.mode(SaveMode.Append).buildForBatch()
-          assert(write != null)
-          WriteToDataSourceV2Exec(write, planLater(query)) :: Nil
-        case _ => throw new AnalysisException(s"data source ${r.name} does not support SaveMode")
-      }
+      AppendDataExec(
+        r.table.asBatchWritable, r.options.toDataSourceOptions, planLater(query)) :: Nil
+
+    case OverwriteByExpression(r: DataSourceV2Relation, deleteExpr, query, _) =>
+      // fail if any filter cannot be converted. correctness depends on removing all matching data.
+      val filters = splitConjunctivePredicates(deleteExpr).map {
+        filter => DataSourceStrategy.translateFilter(deleteExpr).getOrElse(
+          throw new SparkException(s"Cannot translate expression to source filter: $filter"))
```

Comment:
Maybe, `AnalysisException` is clearer than `SparkException`?

Reply:
Conversion to physical plan happens after the analysis is finished, so I think the right exception to use is SparkException.

Reply:
@rdblue .
I know the logic, but we already use `AnalysisException` in this file. It's based on the characteristic of the exception instead of the location. Please see what you removed.
- https://github.com/apache/spark/pull/23606/files#diff-fc4a83061d40a242e7f7b729cb8ee870L153

Reply:
@cloud-fan, what is the correct exception to use during conversion to a physical plan?

Reply:
`AnalysisException` is not currently referenced in `DataSourceV2Strategy`. I also checked `DataSourceStrategy` and that does not throw `AnalysisException` except in the `DataSourceAnalysis` rules that are run by the analyzer.

An alternative is to consider this a failure to produce a plan and return `Nil`. But that would result in an unhelpful assertion failure, which is worse than a specific `SparkException`.

I think this is doing the right thing.

Reply:
Agree that `AnalysisException` is not proper here, as this is not the analysis phase. But `SparkException` is worse. `SparkException` should be used at executor side, which indicates a Spark job is launched.

One thing we can do is, do this conversion at analysis phase. i.e. the logical plan `OverwriteByExpression` should take `Array[Filter]` as well. However, the downside is, for advanced users that touch catalyst rules directly, they are not able to support arbitrary expressions.

So I think `AnalysisException` is OK here, as this is something we should do at the analysis phase, but we don't for other reasons.

Reply:
I'll change this to `AnalysisException`.